pkg/api/api.go

```diff
@@ -62,154 +28,77 @@ type apiImpl struct {
-}
-
-func NewApi(
-	id string,
-	apiPath string,
-	propertyNameOfGetAllResponse string,
-	isSingleConfigurationApi bool,
-	isNonUniqueNameApi bool,
-	isDeprecatedBy string,
-	skipDownload bool,
-) Api {
-
-	// TODO log warning if the user tries to create an API with a id not present in map above
-	// This means that a user runs monaco with an untested api
-
-	return &apiImpl{
-		id:                           id,
-		apiPath:                      apiPath,
-		propertyNameOfGetAllResponse: propertyNameOfGetAllResponse,
-		isSingleConfigurationApi:     isSingleConfigurationApi,
-		isNonUniqueNameApi:           isNonUniqueNameApi,
-		deprecatedBy:                 isDeprecatedBy,
-		skipDownload:                 skipDownload,
-	}
-}
-
-func (a *apiImpl) GetUrl(environmentUrl string) string {
+func (a *API) GetUrl(environmentUrl string) string {
 	return environmentUrl + a.apiPath
 }
 
-func (a *apiImpl) GetId() string {
+func (a *API) GetId() string {
 	return a.id
 }
 
-func (a *apiImpl) GetPropertyNameOfGetAllResponse() string {
+func (a *API) GetPropertyNameOfGetAllResponse() string {
 	return a.propertyNameOfGetAllResponse
 }
 
-func (a *apiImpl) IsStandardApi() bool {
+func (a *API) IsStandardApi() bool {
 	return a.propertyNameOfGetAllResponse == standardApiPropertyNameOfGetAllResponse
 }
 
 // Single configuration APIs are those APIs that configure an environment global setting.
 // Such settings require additional handling and can't be deleted.
-func (a *apiImpl) IsSingleConfigurationApi() bool {
+func (a *API) IsSingleConfigurationApi() bool {
 	return a.isSingleConfigurationApi
 }
 
 // Non unique name APIs are those APIs that don't work with an environment wide unique id.
 // For such APIs, the name attribute can't be used as a id (Monaco default behavior), hence
 // such APIs require additional handling.
-func (a *apiImpl) IsNonUniqueNameApi() bool {
+func (a *API) IsNonUniqueNameApi() bool {
 	return a.isNonUniqueNameApi
 }
 
-func (a *apiImpl) DeprecatedBy() string {
+func (a *API) DeprecatedBy() string {
 	return a.deprecatedBy
 }
 
-func (a *apiImpl) ShouldSkipDownload() bool {
+// ShouldSkipDownload indicates whether an API should be downloaded or not.
+//
+// Some APIs are not re-uploadable by design, either as they require hidden credentials,
+// or if they require a special format, e.g. a zip file.
+//
+// Those configs include all configs handling credentials, as well as the extension-API.
+func (a *API) ShouldSkipDownload() bool {
 	return a.skipDownload
 }
 
-// Contains return true iff requested API is part of particular ApiMap
-func (m ApiMap) Contains(api string) bool {
-	_, ok := m[api]
-	return ok
+// newAPI create new API as a copy of template with a given name
+func newAPI(name string, template *API) *API {
+	n := *template // if API structure is contains pointers or substructure, perform deep copy!!!
+	n.id = name
+	n.applyRules(nameOfGetAllResponse, singleConfigurationApi) //order of applying roles is important!!!
```

Comment:
It's unclear to me what are "rulses" that need to be applied?
as it seems you just check for different combination of set values on the API values.
Couldnt that just be hardcoded in the actual configEndpoints values?